src/arch/aarch64/kernel/processor.rs

```diff
@@ -7,6 +8,128 @@ use hermit_sync::{Lazy, OnceCell, without_interrupts};
 
 use crate::env;
 
+/// Current FPU state. Saved at context switch when changed.
+///
+/// AArch64 mandates 32 NEON SIMD registers, which are named v0-v32.
+/// Only the lower 64 bits of v8-v15 must be saved (the d parts).
+///
+/// See the ABI documentation for aarch64 on this topic:
+/// <https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#612simd-and-floating-point-registers>
+///
+/// FPCR is the floating point control register and controls things like NaN
+/// propagation, FPSR contains info like carry condition over overflow
+/// condition. These are callee-saved bits.
+#[derive(Clone, Copy, Debug)]
+pub struct FPUState {
+	/// d8 register
+	d8: u64,
+	/// d9 register
+	d9: u64,
+	/// d10 register
+	d10: u64,
+	/// d11 register
+	d11: u64,
+	/// d12 register
+	d12: u64,
+	/// d13 register
+	d13: u64,
+	/// d14 register
+	d14: u64,
+	/// d15 register
+	d15: u64,
+	/// fpcr register.
+	fpcr: u64,
+	/// fpsr register.
+	fpsr: u64,
+}
+
+impl FPUState {
+	pub fn new() -> Self {
+		Self {
+			d8: 0,
+			d9: 0,
+			d10: 0,
+			d11: 0,
+			d12: 0,
+			d13: 0,
+			d14: 0,
+			d15: 0,
+			fpcr: 0,
+			fpsr: 0,
+		}
+	}
+
+	pub fn restore(&self) {
+		trace!("Restore FPUState at {self:p}");
+
+		unsafe {
+			asm!(
+				".arch_extension fp",
+				"ldr d8, [{fpu_state}, {off_d8}]",
+				"ldr d9, [{fpu_state}, {off_d9}]",
+				"ldr d10, [{fpu_state}, {off_d10}]",
+				"ldr d11, [{fpu_state}, {off_d11}]",
+				"ldr d12, [{fpu_state}, {off_d12}]",
+				"ldr d13, [{fpu_state}, {off_d13}]",
+				"ldr d14, [{fpu_state}, {off_d14}]",
+				"ldr d15, [{fpu_state}, {off_d15}]",
+				"msr fpcr, {intermediate}",
+				"ldr {intermediate}, [{fpu_state}, {off_fpcr}]",
+				"msr fpsr, {intermediate}",
+				"ldr {intermediate}, [{fpu_state}, {off_fpsr}]",
```

Comment:
`mrs <Xt>, fpcr` stores fpcr in Xt.
`msr fcpr, <Xt>` stores Xt in fpcr.

You got this mixed up here.